src/a_star.py

```diff
@@ -39,13 +40,244 @@ def return_path(current_node):
         current = current.parent
     return path[::-1]  # Return reversed path
 
+def add_floor_to_steps(steps, floor):
+    new_steps = []
+    for step in steps:
+        if step is not None:
+            new_step = [(floor, x, y) for x, y in step]
+            new_steps.append(new_step)
+        else:
+            new_steps.append(None)
+    return new_steps
 
-def astar(
+def astar_partitioned(
     maze: np.ndarray,
     start: Tuple[int, int, int],
     end: Tuple[int, int, int],
     allow_diagonal_movement: bool = False
-) -> Tuple[List[List[Tuple[int, int, int]]], List[List[Tuple[int, int, int]]], List[List[Tuple[int, int, int]]]]:
+) -> Tuple[List[List[Tuple[int, int, int]]], List[List[Tuple[int, int, int]]]]:
+    if start[0] == end[0]:
+        path_step, border_step = astar_euclidean(maze, start, end, allow_diagonal_movement)
+        return path_step, border_step
+    
+    f = open("res/stairs.json")
+    stairs = json.load(f)
+    f.close()
+
+    start_euclidean_distances = []
+    end_euclidean_distances = []
+
+    start_2d = (start[1], start[2])
+    end_2d = (end[1], end[2])
+
+    for f, x, y in stairs:
+        if f == start[0]:
+            distance = np.sqrt((x - start[1]) ** 2 + (y - start[2]) ** 2)
+            coordinates = (x, y)
+            start_euclidean_distances.append((coordinates, distance))
+        elif f == end[0]:
+            distance = np.sqrt((x - end[1]) ** 2 + (y - end[2]) ** 2)
+            coordinates = (x, y)
+            end_euclidean_distances.append((coordinates, distance))
+
+    start_euclidean_distances.sort(key=lambda x: x[1])
+    end_euclidean_distances.sort(key=lambda x: x[1])
+
+    paths_in_start_floor = []
+    paths_in_end_floor = []
+
+    path_step = []
+    border_step = []
+
+    i = 0
+    stop = 2
+    while 1:
+        if i < len(start_euclidean_distances):
+            destiny = start_euclidean_distances[i][0]
+            paths, borders = astar_2d(maze[start[0]], start_2d, destiny, allow_diagonal_movement)
+
+            paths = add_floor_to_steps(paths, start[0])
+            borders = add_floor_to_steps(borders, start[0])
+
+            path_step += paths
+            border_step += borders
+            path = paths[-1]
+            if path is not None:
+                end_calculated_destinations = [x[0] for x in paths_in_end_floor]
+                if destiny in end_calculated_destinations:
+                    end_path = paths_in_end_floor[end_calculated_destinations.index(destiny)][1]
```

Comment:
would it be possible to reorganize this code to use more "continues" in order to lower the amount of indentation and ease understanding?

Reply:
For instance, the major if-else in this block could be swapped with `if i >= len(start_euclidean_distances)` and a continue added at the end, which would make the code easier to reason about

Reply:
the astar_partitioned code is utter trash (one of the worst piece of computer instructions i've ever done in my life)

i'm going to make it more readable

Reply:
fixed it, hopefully is better